internal/pkg/agent/application/upgrade/artifact/download/http/progress_observer.go

```diff
@@ -95,3 +97,38 @@ func (lpObs *loggingProgressObserver) ReportFailed(sourceURI string, timePast ti
 		lpObs.log.Warnf(msg, args...)
 	}
 }
+
+type detailsProgressObserver struct {
+	upgradeDetails *details.Details
+	mu             sync.RWMutex
+}
+
+func newDetailsProgressObserver(upgradeDetails *details.Details) *detailsProgressObserver {
+	upgradeDetails.SetState(details.StateDownloading)
+	return &detailsProgressObserver{
+		upgradeDetails: upgradeDetails,
+	}
+}
+
+func (dpObs *detailsProgressObserver) Report(sourceURI string, timePast time.Duration, downloadedBytes, totalBytes, percentComplete, downloadRate float64) {
+	dpObs.mu.Lock()
+	defer dpObs.mu.Unlock()
+
+	dpObs.upgradeDetails.Metadata.DownloadPercent = percentComplete
+	dpObs.upgradeDetails.NotifyObservers()
+}
+
+func (dpObs *detailsProgressObserver) ReportCompleted(sourceURI string, timePast time.Duration, downloadRate float64) {
+	dpObs.mu.Lock()
+	defer dpObs.mu.Unlock()
+
+	dpObs.upgradeDetails.Metadata.DownloadPercent = 1
+	dpObs.upgradeDetails.NotifyObservers()
+}
```

Comment:
No there is no specific state for a completed download.  In general, we don't have "completed" states for every step, otherwise it would ~2x the number of upgrade states.  We might introduce them in the future if we see a need for it but, at least to start out, we're just recording the states representing "ongoing" upgrade steps.

Reply:
Ok, so `Failure` has a dedicated state for which we have to save previous state information, while for `Success` we rely on convention of `Downloading` and percentage being 100% ?
Interesting choice, I would have preferred a final state for download complete and a clear process outcome (separate field which can be success or failure) but I guess that varies from person to person